include/eve/module/math/regular/exp.hpp

```diff
@@ -83,7 +94,77 @@ namespace eve
 //!        @godbolt{doc/math/masked/exp.cpp}
 //!  @}
 //================================================================================================
-EVE_MAKE_CALLABLE(exp_, exp);
-}
+  inline constexpr auto exp = functor<exp_t>;
 
-#include <eve/module/math/regular/impl/exp.hpp>
+  namespace detail
+  {
+    template<floating_ordered_value T, callable_options O>
+    EVE_FORCEINLINE constexpr T
+    exp_(EVE_REQUIRES(cpu_), O const & o, T x) noexcept
+    {
+      auto isnan = is_nan(x);
+      auto    minlogval = []() {
+        if constexpr(O::contains(pedantic2) && eve::platform::supports_denormals)
+        return minlogdenormal(eve::as<T>());
+        else
+          return minlog(eve::as<T>());
+      };
+      if constexpr( scalar_value<T> )
+      {
+        if constexpr( has_emulated_abi_v<wide<T>> )
+        {
+          return (x <= minlogval()) ? T(0) : (x >= eve::maxlog(as(x))) ? inf(as(x)) : std::exp(x);
+        }
+        else
+        {
+          wide<T> xx(x);
+          return exp(xx).get(0);
+        }
+      }
+      else if constexpr( has_native_abi_v<T> )
+      {
+        using elt_t       = element_type_t<T>;
+        const T Log_2hi   = ieee_constant<0x1.6300000p-1f, 0x1.62e42fee00000p-1>(eve::as<T>{});
+        const T Log_2lo   = ieee_constant<-0x1.bd01060p-13f, 0x1.a39ef35793c76p-33>(eve::as<T>{});
+        const T Invlog_2  = ieee_constant<0x1.7154760p+0f, 0x1.71547652b82fep+0>(eve::as<T>{});
+        auto    xltminlog = x <= minlogval();
+        auto    xgemaxlog = x >= maxlog(eve::as(x));
+        if constexpr( scalar_value<T> )
+        {
+          if( isnan ) return x;
+          if( xgemaxlog ) return inf(eve::as(x));
+          if( xltminlog ) return zero(eve::as(x));
+        }
+        auto c = nearest(Invlog_2 * x);
+        auto k = c;
+        x      = fnma(c, Log_2hi, x); // x-c*L
+        if constexpr( std::is_same_v<elt_t, float> )
+        {
+          x      = fnma(c, Log_2lo, x);
+          auto y = eve::reverse_horner(x, T(0x1.000000p-1f), T(0x1.55534ap-3f), T(0x1.5552aep-5f)
+                                      , T(0x1.131b16p-7f), T(0x1.6ef19ep-10f));;
+          c      = inc(fma(y, sqr(x), x));
+        }
+        else if constexpr( std::is_same_v<elt_t, double> )
```

Comment:
why will we be here? If the previous if is `native abi` already? What are we trying to achieve?